File: backend/app/schemas/race.py
"""
Pydantic schemas for race data
"""

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum


class CompoundType(str, Enum):
    SOFT = "SOFT"
    MEDIUM = "MEDIUM"
    HARD = "HARD"
    INTERMEDIATE = "INTERMEDIATE"
    WET = "WET"


class StintType(str, Enum):
    OPENING = "OPENING"
    MID = "MID"
    CLOSING = "CLOSING"


class TrackStatus(str, Enum):
    GREEN = "GREEN"
    YELLOW = "YELLOW"
    SC = "SC"
    VSC = "VSC"
    RED = "RED"


class SeasonResponse(BaseModel):
    """Season response schema"""
    id: int
    year: int

    class Config:
        from_attributes = True


class RaceBase(BaseModel):
    """Base race schema"""
    grand_prix: str
    gp_slug: str
    event_date: datetime
    round_number: Optional[int] = None
    circuit_name: str
    circuit_short: Optional[str] = None
    circuit_country: Optional[str] = None
    track_length_km: Optional[float] = None
    altitude_m: Optional[float] = None
    circuit_type: Optional[str] = None


class RaceResponse(RaceBase):
    """Race response schema"""
    id: int
    season_id: int

    class Config:
        from_attributes = True


class LapBase(BaseModel):
    """Base lap schema"""
    driver: str
    driver_number: Optional[int] = None
    team: str
    lap_number: int
    lap_time: Optional[str] = None
    lap_time_seconds: Optional[float] = None

    # Sector times
    sector1_time_seconds: Optional[float] = None
    sector2_time_seconds: Optional[float] = None
    sector3_time_seconds: Optional[float] = None

    # Speed traps
    speed_i1: Optional[float] = None
    speed_i2: Optional[float] = None
    speed_fl: Optional[float] = None
    speed_st: Optional[float] = None

    # Tyre info
    compound: Optional[CompoundType] = None
    tyre_life: Optional[int] = None
    is_fresh_tyre: bool = False

    # Stint info
    stint: Optional[int] = None
    stint_type: Optional[StintType] = None

    # Position
    position: Optional[int] = None

    # Track status
    track_status: Optional[TrackStatus] = None

    # Flags
    is_personal_best: bool = False
    is_valid_lap: bool = True
    is_accurate: bool = True


class LapResponse(LapBase):
    """Lap response schema"""
    id: int
    race_id: int

    class Config:
        from_attributes = True


class WeatherDataResponse(BaseModel):
    """Weather data response schema"""
    id: int
    race_id: int
    lap_number: Optional[int] = None
    air_temp: Optional[float] = None
    track_temp: Optional[float] = None
    humidity: Optional[float] = None
    pressure: Optional[float] = None
    wind_speed: Optional[float] = None
    wind_direction: Optional[float] = None
    rainfall: bool = False
    air_density: Optional[float] = None

    class Config:
        from_attributes = True


class RaceDataQuery(BaseModel):
    """Query parameters for race data"""
    season: Optional[int] = None
    grand_prix: Optional[str] = None
    drivers: Optional[List[str]] = Field(default=None, description="List of driver codes")
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=100, ge=1, le=1000)


class RaceStatistics(BaseModel):
    """Race statistics summary"""
    driver: str
    team: str
    total_laps: int
    avg_lap_time: float
    fastest_lap: float
    pit_stops: int
    positions_gained: int
    final_position: int


class RaceDataResponse(BaseModel):
    """Complete race data response"""
    race: RaceResponse
    laps: List[LapResponse]
    weather: List[WeatherDataResponse]
    statistics: List[RaceStatistics]
    total_laps: int
    page: int
    page_size: int

File: frontend/lib/types/race.ts
/**
 * TypeScript types for race data
 */

export enum CompoundType {
  SOFT = 'SOFT',
  MEDIUM = 'MEDIUM',
  HARD = 'HARD',
  INTERMEDIATE = 'INTERMEDIATE',
  WET = 'WET',
}

export enum StintType {
  OPENING = 'OPENING',
  MID = 'MID',
  CLOSING = 'CLOSING',
}

export enum TrackStatus {
  GREEN = 'GREEN',
  YELLOW = 'YELLOW',
  SC = 'SC',
  VSC = 'VSC',
  RED = 'RED',
}

export interface Season {
  id: number;
  year: number;
}

export interface Race {
  id: number;
  season_id: number;
  grand_prix: string;
  gp_slug: string;
  event_date: string;
  round_number: number | null;
  circuit_name: string;
  circuit_short: string | null;
  circuit_country: string | null;
  track_length_km: number | null;
  altitude_m: number | null;
  circuit_type: string | null;
}

export interface Lap {
  id: number;
  race_id: number;
  driver: string;
  driver_number: number | null;
  team: string;
  lap_number: number;
  lap_time: string | null;
  lap_time_seconds: number | null;
  sector1_time_seconds: number | null;
  sector2_time_seconds: number | null;
  sector3_time_seconds: number | null;
  sector1_pct: number | null;
  sector2_pct: number | null;
  sector3_pct: number | null;
  speed_i1: number | null;
  speed_i2: number | null;
  speed_fl: number | null;
  speed_st: number | null;
  compound: CompoundType | null;
  tyre_life: number | null;
  is_fresh_tyre: boolean;
  stint: number | null;
  stint_type: StintType | null;
  position: number | null;
  track_status: TrackStatus | null;
  is_personal_best: boolean;
  is_valid_lap: boolean;
  is_accurate: boolean;
  delta_to_fastest_lap: number | null;
  is_pit_out_lap?: boolean;
  is_pit_in_lap?: boolean;
}

export interface WeatherData {
  id: number;
  race_id: number;
  lap_number: number | null;
  air_temp: number | null;
  track_temp: number | null;
  humidity: number | null;
  pressure: number | null;
  wind_speed: number | null;
  wind_direction: number | null;
  rainfall: boolean;
  air_density: number | null;
}

export interface RaceStatistics {
  driver: string;
  team: string;
  total_laps: number;
  avg_lap_time: number;
  fastest_lap: number;
  pit_stops: number;
  positions_gained: number;
  final_position: number;
}

export interface RaceDataResponse {
  race: Race;
  laps: Lap[];
  weather: WeatherData[];
  statistics: RaceStatistics[];
  total_laps: number;
  page: number;
  page_size: number;
}

export interface CircuitLayout {
  circuit_name: string;
  circuit_country: string;
  track_length_km: number | null;
  altitude_m: number | null;
  coordinates: {
    x: number;
    y: number;
    speed: number;
    distance: number;
  }[];
  fastest_lap_time: string | null;
  fastest_lap_driver: string | null;
}
